desc/integrals/bounce_utils.py

```diff
@@ -0,0 +1,819 @@
+"""Utilities and functional programming interface for bounce integrals."""
+
+from interpax import PPoly
+from matplotlib import pyplot as plt
+
+from desc.backend import imap, jnp
+from desc.backend import softmax as softargmax
+from desc.integrals.basis import _add2legend, _in_epigraph_and, _plot_intersect
+from desc.integrals.interp_utils import (
+    interp1d_Hermite_vec,
+    interp1d_vec,
+    poly_root,
+    polyval_vec,
+)
+from desc.integrals.quad_utils import (
+    bijection_from_disc,
+    composite_linspace,
+    grad_bijection_from_disc,
+)
+from desc.utils import atleast_3d_mid, errorif, setdefault, take_mask
+
+
+def get_pitch(min_B, max_B, num, relative_shift=1e-6):
+    """Return 1/λ values uniformly spaced between ``min_B`` and ``max_B``.
+
+    Parameters
+    ----------
+    min_B : jnp.ndarray
+        Minimum |B| value.
+    max_B : jnp.ndarray
+        Maximum |B| value.
+    num : int
+        Number of values, not including endpoints.
+    relative_shift : float
+        Relative amount to shift maxima down and minima up to avoid floating point
+        errors in downstream routines.
+
+    Returns
+    -------
+    pitch : jnp.ndarray
+        Shape (num + 2, *min_B.shape).
+        1/λ values. Note ``pitch`` = 1/λ ~ E/μ = energy / magnetic moment.
+
+    """
+    # Floating point error impedes consistent detection of bounce points riding
+    # extrema. Shift values slightly to resolve this issue.
+    min_B = (1 + relative_shift) * min_B
+    max_B = (1 - relative_shift) * max_B
+    # Samples should be uniformly spaced in |B| and not λ (GitHub issue #1228).
```

Comment:
another maybe dumb question:

We use gaussian quadrature for the integrals along zeta, would there be any reason to use something similar for the integrals over the pitch angle?

Reply:
FYI right now we do simpson's for effective ripple which is 3rd order piecewise Gaussian quadrature, and trapezoid for $\Gamma_c$.